processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -240,22 +257,46 @@ public int getLastRowIndex()
     return indexIncrement.get() - 1;
   }
 
-  private void factorizeAggs(
+  /**
+   * Creates aggregators for the given aggregator factories.
+   *
+   * @return Total initial size in bytes required by all the aggregators.
+   * This value is non-zero only when {@link #useMaxMemoryEstimates} is false.
+   */
+  private long factorizeAggs(
       AggregatorFactory[] metrics,
       Aggregator[] aggs,
       ThreadLocal<InputRow> rowContainer,
       InputRow row
   )
   {
+    long totalInitialSizeBytes = 0L;
     rowContainer.set(row);
     for (int i = 0; i < metrics.length; i++) {
       final AggregatorFactory agg = metrics[i];
-      aggs[i] = agg.factorize(selectors.get(agg.getName()));
+
+      if (useMaxMemoryEstimates) {
+        aggs[i] = agg.factorize(selectors.get(agg.getName()));
+      } else {
+        SizedAggregator sizedAggregator =
+            agg.factorizeSized(selectors.get(agg.getName()));
+        aggs[i] = sizedAggregator.getAggregator();
+        totalInitialSizeBytes += sizedAggregator.getInitialSizeBytes();
+      }
     }
     rowContainer.set(null);
+
+    return useMaxMemoryEstimates ? 0 : totalInitialSizeBytes;
```

Comment:
You've got the same check in multiple places.  I don't think you need the one here as `totalInitialSizeBytes` will be 0 in the `useMaxMemoryEstimates` case anyway.  But if you really want to ensure that the two code paths don't mix, move the `if(useMaxMemoryEstimates)` check to the top level and write the loop twice.  Once in a block that always returns `0` and another in a block that returns an aggregated sum.

Reply:
Fixed.